### ISSUE TITLE
Connect via MQTT and subscribe to all telemetry messages

### ISSUE THREAD

**Issue by Poquaruse** [thingsboard/thingsboard#318]
Hi,

I'm more and more surprised by the features that thingsboard offers. So many great ideas! 👍 

We're thinking about implementing an energy management system based on thingsboard. We'd implement an external service which gets data, does its optimization magic and then sends rpcs to the devices which need to be controlled. thingsboard would serve as visualization and control instance. However, I struggle to find out whether (and how) it is possible for a device/a service to receive any (or a defined set of) MQTT message(s) which is/are send to thingsboard. The problem is that we can't do the optimization inside thingsboard but need all available data from in there...

I'd be happy if you could point me into the right direction here... 

Thanks and best regards!

**Comment by ashvayka** [comment]
Hi @Poquaruse . Can you attach a flow diagram, because it's hard to parse the flow from your text.

**Comment by Poquaruse** [comment]
Hi @ashvayka,

thank you for your answer. I'm currently not on my working laptop and therefore it'll take some time until I can attach a flow diagram. Sorry!

The idea is quite simple: Is there a way for a service (an external process connected via MQTT) to receive updates from all devices (telemetry, rpcs, attributes) via MQTT? 

Some kind of "replay mechanism" such that this external service would get messages like "Device XYZ has send the following telemetry data" and "There is an RPC for Device ABC with the following params" and so on. So an external device should receive a copy of all messages that are sent to the thingsboard instance...

Thanks and best regards!

**Comment by z4nD4R** [comment]
@Poquaruse ,

let me know when you figure it out. As I did test: In theory mqtt broker should send the message to the clients which are subscribed to particular tokens. So when you have one client "posting" to TB via some topics and you will also subscribe with another client to this topic you SHOULD receive the message. Howerver that's theory. TB broker is not working like that.

I have similar scenario where I need one particular node would receive all mqtt messages desitined to TB :) Only method I could achieve that was 2 message published with client

1. One to TB (as original)
2. Second directly to device via mqtt (there is mechanism described on documentation how 2 devices can communicate to each other)

Hope it helps

Y

**Comment by Poquaruse** [comment]
As far as I know, thingsboard implements its own MQTT broker. And that broker does not send messages to all those who subscribe to a certain topic. So as you described, this way does not work unfortunately.

To be honest, I thought about using a second MQTT broker on my network and publishing to thingsboard as well as to that second broker in parallel. But that seems unnecessary and hacky.

Could you please point me to that part of the documentation? I only found rpcs but that is not what they are meant for... Perhaps I've overlooked something?

**Comment by z4nD4R** [comment]
Hi @Poquaruse ,

yeah it is a bit tricky to find it there but here is the link: https://thingsboard.io/docs/reference/plugins/messaging/

Check the `mqtt-send-msg.js`.

I'm using this approach where 1 msg is going to `TB` and secondone is going to "control node" via Messaging capapilities ;) 

Y

**Comment by ikulikov** [comment]
Closing due to inactivity.

**Comment by Poquaruse** [comment]
Hi @ikulikov,

I'm sorry if this issued appeared to be inactive. However, the main problem still isn't solved for me. Furthermore, I guess @z4nD4R might also still be interested in this question. Perhaps you could reopen this issue again, please?

Is there a way to send duplicates of all messages to a device so that this device knows "everything"? (Detailed explanation above...)

Thanks and best regards

**Comment by PopoTheSapien** [comment]
@Poquaruse I had some other MQTT requirements that I couldn't use the TB broker for.  Installing mosquito alongside TB and running one of the two on a different port may be a slightly cleaner workaround.  Works well for me...

**Comment by abhijeetpandey** [comment]
@Poquaruse @ashvayka I face the same problem I want to subscribe to the telemetry data via telemetry topic, to which one device is sending the data. Unfortunately, when I subscribe to telemetry topic, no message is received. It just stalls and keeps waiting.

**Comment by nitinware** [comment]
@Poquaruse @ashvayka @abhijeetpandey - Were you able to resolve the issue, as I also want to subscribe to the telemetry topic and do some processing on telemetry data.

**Comment by pavellzubkov** [comment]
would be fine if tb gateway might connect to TB embedded mqtt broker.

**Comment by psp316r** [comment]
Even I am facing issue , I cannot get the telemetry data using MQTT , Unfortunately, when I subscribed to telemetry topics, no message is received. It just  keeps waiting.

**Comment by Wanderzhd** [comment]
I am facing the same issue as @psp316r described above. 

Any solution?

**Comment by SuneethaYamani** [comment]
I am facing same issue

**Comment by psp316r** [comment]
> 
> 
> I am facing same issue

I got a Solution for Subscribe and Publish which you can do through the Attributes, instead of Telemetry, use Attributes to publish and subscribe that will work, I have written the code for it, but need to check where it is !!!

**Comment by SuneethaYamani** [comment]
@psp316r  for telemetry we have API to fetch data passing startTS and endTS to show analytics,if I upload data to attributes i didn't find any API to show analytics..is there any api to fetch attributes passing startTS and endTS?

**Comment by SuneethaYamani** [comment]
> > I am facing same issue
> 
> I got a Solution for Subscribe and Publish which you can do through the Attributes, instead of Telemetry, use Attributes to publish and subscribe that will work, I have written the code for it, but need to check where it is !!!

"If I use this I can get only current attributes",I want to show analytics also

**Comment by falsovsky** [comment]
I also found weird that subscribe only works for Attributes and not for Telemetry.

**Comment by bioshock2k** [comment]
> > I am facing same issue
> 
> I got a Solution for Subscribe and Publish which you can do through the Attributes, instead of Telemetry, use Attributes to publish and subscribe that will work, I have written the code for it, but need to check where it is !!!

Would like to see that code 🍡

**Comment by andreiUshenko** [comment]
Same issue, for self remote BI, required to call for two types of db.

**Comment by OdayKabaha** [comment]
I have the same issue. Any solutions ?

**Comment by srujanavajrala** [comment]
> > I am facing same issue
> 
> I got a Solution for Subscribe and Publish which you can do through the Attributes, instead of Telemetry, use Attributes to publish and subscribe that will work, I have written the code for it, but need to check where it is !!!

**Comment by srujanavajrala** [comment]
Can you share the code please

**Comment by Vishalbuu** [comment]
I also has same issues any solution?

**Comment by OdayKabaha** [comment]
@Vishalbuu 
IoT Gateway. Create your devices with the Gateway then you can publish and subscribe to the topics. 

You can also use the REST interface to get the telemetry data regularly using REST-Polling. 

I hope it was helpful.

**Comment by Vishalbuu** [comment]
It's still not getting My output
for  mosquitto_sub -d -h "192.168.1.21" -t "v1/gateway/attributes/" -u "76xd8q19DRlEguxuXJMt" -q 2

Client mosq-o4oD1jKCIbS7tohV4i sending CONNECT
Client mosq-o4oD1jKCIbS7tohV4i received CONNACK (0)
Client mosq-o4oD1jKCIbS7tohV4i sending SUBSCRIBE (Mid: 202, Topic: v1/gateway/attributes/request, QoS: 2, Options: 0x00)
Client mosq-o4oD1jKCIbS7tohV4i received SUBACK
Subscribed (mid: 202): 128
Client mosq-o4oD1jKCIbS7tohV4i sending CONNECT
Client mosq-o4oD1jKCIbS7tohV4i received CONNACK (0)
Client mosq-o4oD1jKCIbS7tohV4i sending SUBSCRIBE (Mid: 203, Topic: v1/gateway/attributes/request, QoS: 2, Options: 0x00)
Client mosq-o4oD1jKCIbS7tohV4i received SUBACK
Subscribed (mid: 203): 128
......
But I am not getting expected output{"":""}

**Comment by jembishop** [comment]
I would also like to add this feature would be very useful in my use case.
